src/fp.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) 2021 The Authors
+// SPDX-License-Identifier: MPL-2.0
+
+use rand::prelude::*;
+use rand::Rng;
+
+// Parameters of a finite field with prime p < 2^126. Field elements are represented as u128's.
+#[derive(Debug)]
+pub struct Fp {
+    pub p: u128,  // the prime modulus p
+    pub p2: u128, // p * 2
+    pub mu: u64,  // mu = -p^(-1) mod 2^64
+    pub r2: u128, // r2 = (2^128)^2 mod p
+}
+
+impl Fp {
```

Comment:
Could there be an associated function `Fp::new(p: u128)` that computes `p2`, `mu` and `r2`, reducing the risk that a caller will compute them wrong?

edit: I see there is such a function implemented in the tests. Were you trying to avoid pulling `BigNum` into the regular dependencies?

Reply:
It seems sensible to me to avoid adding new dependencies to the library wherever possible, in order to keep it as small possible. Though TBH I don't know enough about Rust internals to know how much this actually matters.

I'd be happy to implement `Fp::new(p : u128)`. Note that this would actually add `modinverse` as a dependency in addition to `num-bigint`. (The latter doesn't do modular inverse AFACIT.)

Reply:
Take a look on Alg.3, it's easy to derive `mu` so you can remove such a dependency.
https://eprint.iacr.org/2017/1057.pdf

Reply:
Neat! This leaves the `num-bigint` dependency. Anyway we can get rid of this with some mathmagic?

Reply:
One option would be to provide a method `FieldParameters::validate(&self) -> bool`, tagged with `#[cfg(test)]`, which would take a `FieldParameters` and return true if the four values check out against each other. Then `FieldParameters::validate` could be used in the `Field` module's own tests to verify that `SMALL_FP` is valid.

Reply:
That's a great idea! Working..

Reply:
Done.